helix-core/src/main/java/org/apache/helix/tools/ClusterSetup.java

```diff
@@ -176,6 +179,31 @@ public void close() {
     }
   }
 
+  public void addCluster(String clusterName, boolean overwritePrevious, CloudConfig cloudConfig)
+      throws HelixException {
+    if (cloudConfig != null) {
+      if (cloudConfig.isCloudEnabled()) {
+        String cloudID = cloudConfig.getCloudID();
+        String cloudProvider = cloudConfig.getCloudProvider();
+        List<String> cloudInfoSources = cloudConfig.getCloudInfoSources();
+        String cloudProcessorName = cloudConfig.getCloudInfoProcessorName();
+
+        // Validate the cloudConfig in the builder
+        // If it is invalid, do not even create a cluster
+        CloudConfig.Builder builder = new CloudConfig.Builder(clusterName, true,
+            CloudProvider.valueOf(cloudProvider), cloudID, cloudInfoSources, cloudProcessorName);
+
+        CloudConfig validatedCloudConfig = builder.build();
+        addCluster(clusterName, overwritePrevious);
+        _admin.addCloudConfig(clusterName, validatedCloudConfig);
+      } else {
+        addCluster(clusterName, overwritePrevious);
+      }
+    } else {
+      throw new HelixException("Cannot add the cluster, cloudConfig is null!");
+    }
+  }
+
   public void addCluster(String clusterName, boolean overwritePrevious) {
     _admin.addCluster(clusterName, overwritePrevious);
```

Comment:
In general, usually, we do this in a reverse way. The method with less parameters should call the method with more parameters. I think that is why you have the awkward condition check in the method above.

Reply:
Good suggestion. Fixed.